as.c

```diff
@@ -32,11 +33,338 @@
+    InternetSetOptionA(hRequest, INTERNET_OPTION_PASSWORD, (LPVOID)password, (DWORD)strlen(password));
+
+    /* handle cert errors */
+    /* https://www.betaarchive.com/wiki/index.php/Microsoft_KB_Archive/182888 */
+    if (!HttpSendRequestW(hRequest, NULL, 0, NULL, 0)) {
+        DWORD err = GetLastError();
+        if ((err == ERROR_INTERNET_INVALID_CA) ||
+            (err == ERROR_INTERNET_SEC_CERT_CN_INVALID) ||
+            (err == ERROR_INTERNET_SEC_CERT_DATE_INVALID) ||
+            (err == ERROR_INTERNET_SEC_CERT_REV_FAILED)) {
+
+            /* ask user what to do and modify options if needed */
+            DWORD dlg_result = InternetErrorDlg(hWnd, hRequest,
+                err,
+                FLAGS_ERROR_UI_FILTER_FOR_ERRORS |
+                FLAGS_ERROR_UI_FLAGS_GENERATE_DATA |
+                FLAGS_ERROR_UI_FLAGS_CHANGE_OPTIONS,
+                NULL);
+
+            if (dlg_result == ERROR_SUCCESS) {
+                /* for unknown reasons InternetErrorDlg() doesn't change options for ERROR_INTERNET_SEC_CERT_REV_FAILED,
+                 * despite user is willing to continue, so we have to do it manually */
+                if (err == ERROR_INTERNET_SEC_CERT_REV_FAILED) {
+                    DWORD flags;
+                    DWORD len = sizeof(flags);
+                    InternetQueryOption(hRequest, INTERNET_OPTION_SECURITY_FLAGS, (LPVOID)&flags, &len);
+
+                    flags |= SECURITY_FLAG_IGNORE_REVOCATION;
+                    InternetSetOption(hRequest, INTERNET_OPTION_SECURITY_FLAGS, &flags, sizeof(flags));
+                    goto again;
+                }
+
+                SetCursor(LoadCursorW(0, IDC_WAIT));
+                goto again;
+            }
+            else
+                goto done;
+        }
+
+        ShowWinInetError(hWnd);
+        goto done;
+    }
+
+    /* get http status code */
+    DWORD statusCode = 0;
+    DWORD length = sizeof(DWORD);
+    HttpQueryInfoW(hRequest, HTTP_QUERY_STATUS_CODE | HTTP_QUERY_FLAG_NUMBER, &statusCode, &length, NULL);
+    if (statusCode != 200) {
+        ShowLocalizedMsgEx(MB_OK, hWnd, _T(PACKAGE_NAME), IDS_ERR_URL_IMPORT_PROFILE, statusCode, "HTTP error");
+        goto done;
+    }
+
+    /* download profile content */
+    size_t size = 0;
+    if (!DownloadProfileContent(hWnd, hRequest, &buf, &size))
+        goto done;
+
+    WCHAR name[MAX_PATH] = {0};
+    WCHAR* wbuf = Widen(buf);
+    if (!wbuf) {
+        MessageBoxW(hWnd, L"Failed to convert profile content to wchar", _T(PACKAGE_NAME), MB_OK);
+        goto done;
+    }
+    ExtractProfileName(wbuf, comps.host, name, MAX_PATH);
+    free(wbuf);
+
+    /* save profile content into tmp file */
+    DWORD res = GetTempPathW(MAX_PATH, out_path);
+    if (res == -1 || res > MAX_PATH) {
+        MessageBoxW(hWnd, L"Failed to get TMP path", _T(PACKAGE_NAME), MB_OK);
+        goto done;
+    }
+    wcscat(out_path, name);
```

Comment:
Use wcsncat or our sntprintf_0 to append the file name. Also safer to pass in the size of out_path or ensure at least room for MAX_PATH+1 is available.

Reply:
Changed to `swprintf` and NULL termination.